File: scripts/create-field.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

// Get field name from command line arguments
const fieldName = process.argv[2];

if (!fieldName) {
  console.error('❌ Error: Please provide a field name');
  console.log('Usage: yarn create:field <FieldName>');
  console.log('Example: yarn create:field MyCustomField');
  process.exit(1);
}

// Convert to PascalCase and validate
const pascalCaseName = fieldName.replace(/(^\w|-\w)/g, (match) =>
  match.replace('-', '').toUpperCase(),
);

if (!/^[A-Z][a-zA-Z0-9]*$/.test(pascalCaseName)) {
  console.error(
    '❌ Error: Field name must be in PascalCase (e.g., MyCustomField)',
  );
  process.exit(1);
}

const kebabCaseName = pascalCaseName
  .replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, '$1-$2')
  .toLowerCase();
const componentPath = path.join(
  __dirname,
  '..',
  'formily',
  'antd',
  'src',
  'components',
  pascalCaseName,
);
const schemaPath = path.join(
  __dirname,
  '..',
  'formily',
  'antd',
  'src',
  'schemas',
);

console.log('🚀 Creating field template for: ' + pascalCaseName);

// Create component directory
if (!fs.existsSync(componentPath)) {
  fs.mkdirSync(componentPath, { recursive: true });
}

// Generate component files
const files = {
  // Component index.ts
  'index.ts': "export * from './preview';\n",

  // Component preview.tsx
  'preview.tsx':
    "import React from 'react';\nimport { Input as FormilyInput } from '@formily/antd-v5';\nimport { createBehavior, createResource } from '@thevikalp/designable-core';\nimport { DnFC } from '@thevikalp/designable-react';\nimport { createFieldSchema } from '../Field';\nimport { AllSchemas } from '../../schemas';\nimport { AllLocales } from '../../locales';\n\nexport const " +
    pascalCaseName +
    ': DnFC<React.ComponentProps<typeof FormilyInput>> = FormilyInput;\n\n' +
    pascalCaseName +
    ".Behavior = createBehavior({\n  name: '" +
    pascalCaseName +
    "',\n  extends: ['Field'],\n  selector: (node) => node.props['x-component'] === '" +
    pascalCaseName +
    "',\n  designerProps: {\n    propsSchema: createFieldSchema(AllSchemas." +
    pascalCaseName +
    '),\n  },\n  designerLocales: AllLocales.' +
    pascalCaseName +
    ',\n});\n\n' +
    pascalCaseName +
    ".Resource = createResource({\n  icon: '" +
    pascalCaseName +
    "Source',\n  elements: [\n    {\n      componentName: 'Field',\n      props: {\n        type: 'string',\n        title: '" +
    pascalCaseName +
    "',\n        'x-decorator': 'FormItem',\n        'x-component': '" +
    pascalCaseName +
    "',\n      },\n    },\n  ],\n});\n",
};

// Schema file content
const schemaContent =
  "import { ISchema } from '@formily/react';\n\nexport const " +
  pascalCaseName +
  ": ISchema = {\n  type: 'object',\n  properties: {\n    placeholder: {\n      type: 'string',\n      'x-decorator': 'FormItem',\n      'x-component': 'Input',\n      'x-component-props': {\n        placeholder: 'Enter placeholder text',\n      },\n    },\n    maxLength: {\n      type: 'number',\n      'x-decorator': 'FormItem',\n      'x-component': 'NumberPicker',\n      'x-component-props': {\n        placeholder: 'Maximum length',\n      },\n    },\n    required: {\n      type: 'boolean',\n      'x-decorator': 'FormItem',\n      'x-component': 'Switch',\n      'x-component-props': {\n        defaultChecked: false,\n      },\n    },\n  },\n};\n";

// Create component files
Object.entries(files).forEach(([fileName, content]) => {
  const filePath = path.join(componentPath, fileName);
  fs.writeFileSync(filePath, content);
  console.log(
    '✅ Created: ' + path.relative(path.join(__dirname, '..'), filePath),
  );
});

// Create schema file
const schemaFilePath = path.join(schemaPath, pascalCaseName + '.ts');
fs.writeFileSync(schemaFilePath, schemaContent);
console.log(
  '✅ Created: ' + path.relative(path.join(__dirname, '..'), schemaFilePath),
);

// Update components index.ts
const componentsIndexPath = path.join(
  __dirname,
  '..',
  'formily',
  'antd',
  'src',
  'components',
  'index.ts',
);
let componentsIndex = fs.readFileSync(componentsIndexPath, 'utf8');

// Add export if not already present
if (!componentsIndex.includes("export * from './" + pascalCaseName + "';")) {
  componentsIndex += "\nexport * from './" + pascalCaseName + "';";
  fs.writeFileSync(componentsIndexPath, componentsIndex);
  console.log('✅ Updated: formily/antd/src/components/index.ts');
}

// Update schemas index.ts
const schemasIndexPath = path.join(
  __dirname,
  '..',
  'formily',
  'antd',
  'src',
  'schemas',
  'all.ts',
);
let schemasIndex = fs.readFileSync(schemasIndexPath, 'utf8');

// Add export if not already present
if (!schemasIndex.includes("export * from './" + pascalCaseName + "';")) {
  schemasIndex += "\nexport * from './" + pascalCaseName + "';";
  fs.writeFileSync(schemasIndexPath, schemasIndex);
  console.log('✅ Updated: formily/antd/src/schemas/all.ts');
}

// Update locales
const localesDir = path.join(
  __dirname,
  '..',
  'formily',
  'antd',
  'src',
  'locales',
);

// Create locale file
const localeContent =
  'export const ' +
  pascalCaseName +
  " = {\n  'zh-CN': {\n    title: '" +
  pascalCaseName +
  "',\n    settings: {\n      'x-component-props': {\n        placeholder: '占位提示',\n        maxLength: '最大长度',\n      },\n    },\n  },\n  'en-US': {\n    title: '" +
  pascalCaseName +
  "',\n    settings: {\n      'x-component-props': {\n        placeholder: 'Placeholder',\n        maxLength: 'Max Length',\n      },\n    },\n  },\n  'ko-KR': {\n    title: '" +
  pascalCaseName +
  "',\n    settings: {\n      'x-component-props': {\n        placeholder: '플레이스홀더',\n        maxLength: '최대 길이',\n      },\n    },\n  },\n};\n";

const localeFilePath = path.join(localesDir, pascalCaseName + '.ts');
fs.writeFileSync(localeFilePath, localeContent);
console.log(
  '✅ Created: ' + path.relative(path.join(__dirname, '..'), localeFilePath),
);

// Update locales index.ts
const localesIndexPath = path.join(localesDir, 'all.ts');
let localesIndex = fs.readFileSync(localesIndexPath, 'utf8');

// Add export if not already present
if (!localesIndex.includes("export * from './" + pascalCaseName + "';")) {
  localesIndex += "\nexport * from './" + pascalCaseName + "';";
  fs.writeFileSync(localesIndexPath, localesIndex);
  console.log('✅ Updated: formily/antd/src/locales/all.ts');
}

console.log('\n🎉 Field template created successfully!');
console.log('\n📝 Next steps:');
console.log(
  '1. Edit the generated files in: formily/antd/src/components/' +
    pascalCaseName +
    '/',
);
console.log(
  '2. Customize the schema in: formily/antd/src/schemas/' +
    pascalCaseName +
    '.ts',
);
console.log('3. Update locales in: formily/antd/src/locales/');
console.log('4. Build the project: yarn build');
console.log('5. Test in the example: yarn example:basic');
console.log(
  '\n💡 The field will appear in the "Basic Fields" section of the designer.',
);

File: scripts/create-field.rs
use std::env;
use std::fs;
use std::path::Path;
use std::process;

//main fn
fn main() {
    //vector of strings
    //collecting args from command line
    let args: Vec<String> = env::args().collect();
    //if less than 2 args, print error and exit
    if args.len() < 2 {
        eprintln!("❌ Error: Please provide a field name");
        eprintln!("Usage: {} <FieldName>", args[0]);
        eprintln!("Example: {} MyCustomField", args[0]);
        process::exit(1);
    }

    //getting index 1 arg as field name
    let field_name = &args[1]; //shared reference

    // Convert to PascalCase
    let pascal_case_name = to_pascal_case(field_name); //calling custom function that converts to pascal case
    if !is_valid_pascal_case(&pascal_case_name) {
        //checking if the name is valid pascal case
        eprintln!(
            "❌ Error: Field name must be in PascalCase (e.g., MyCustomField)"
        );
        process::exit(1);
    }

    // Convert to kebab-case
    let _kebab_case_name = to_kebab_case(&pascal_case_name); //calling custom function that converts to kebab case

    // new path object
    let base_dir = Path::new("formily/antd/src");
    // new destination path to components/FieldName
    let component_path = base_dir.join("components").join(&pascal_case_name);
    //new destination path to schemas
    let schema_path = base_dir.join("schemas");

    println!("🚀 Creating field template for: {}", pascal_case_name);

    // Create component directory
    //create all directories if they don't exist
    //if error, print and exit
    if let Err(e) = fs::create_dir_all(&component_path) {
        eprintln!("❌ Error creating directory: {}", e);
        process::exit(1);
    }

    // Generate component files
    //vector of tuples of file name and content
    let files = vec![
        ("index.ts", format!("export * from './preview';\n")),
        ("preview.tsx", generate_preview_content(&pascal_case_name))
    ];

    for (file_name, content) in files {
        let file_path = component_path.join(file_name);
        if let Err(e) = fs::write(&file_path, content) {
            eprintln!("❌ Error writing file {}: {}", file_path.display(), e);
            process::exit(1);
        }
        println!("✅ Created: {}", file_path.display());
    }

    // Create schema file
    let schema_file_path = schema_path.join(format!("{}.ts", pascal_case_name));
    let schema_content = generate_schema_content(&pascal_case_name);
    if let Err(e) = fs::write(&schema_file_path, schema_content) {
        eprintln!("❌ Error writing schema file: {}", e);
        process::exit(1);
    }
    println!("✅ Created: {}", schema_file_path.display());

    // Update components index.ts
    update_index_file(
        &base_dir.join("components/index.ts"),
        &format!("export * from './{}';", pascal_case_name)
    );

    // Update schemas all.ts
    update_index_file(
        &schema_path.join("all.ts"),
        &format!("export * from './{}';", pascal_case_name)
    );

    // Create locale file
    let locales_dir = base_dir.join("locales");
    let locale_file_path = locales_dir.join(format!("{}.ts", pascal_case_name));
    let locale_content = generate_locale_content(&pascal_case_name);
    if let Err(e) = fs::write(&locale_file_path, locale_content) {
        eprintln!("❌ Error writing locale file: {}", e);
        process::exit(1);
    }
    println!("✅ Created: {}", locale_file_path.display());

    // Update locales all.ts
    update_index_file(
        &locales_dir.join("all.ts"),
        &format!("export * from './{}';", pascal_case_name)
    );

    println!("\n🎉 Field template created successfully!");
    println!("\n📝 Next steps:");
    println!("1. Edit the generated files in: formily/antd/src/components/{}/", pascal_case_name);
    println!("2. Customize the schema in: formily/antd/src/schemas/{}.ts", pascal_case_name);
    println!("3. Update locales in: formily/antd/src/locales/");
    println!("4. Build the project: yarn build");
    println!("5. Test in the example: yarn example:basic");
    println!(
        "\n💡 The field will appear in the \"Basic Fields\" section of the designer."
    );
}

fn to_pascal_case(s: &str) -> String {
    s.split('-')
        .map(|word| {
            let mut chars = word.chars();
            match chars.next() {
                None => String::new(),
                Some(first) =>
                    first.to_uppercase().collect::<String>() +
                        chars.as_str().to_lowercase().as_str(),
            }
        })
        .collect()
}

fn to_kebab_case(s: &str) -> String {
    let mut result = String::new();
    for (i, ch) in s.char_indices() {
        if ch.is_uppercase() && i > 0 {
            result.push('-');
        }
        result.push(ch.to_lowercase().next().unwrap());
    }
    result
}

fn is_valid_pascal_case(s: &str) -> bool {
    if s.is_empty() {
        return false;
    }
    let chars: Vec<char> = s.chars().collect();
    if !chars[0].is_uppercase() {
        return false;
    }
    for &ch in &chars[1..] {
        if !ch.is_alphanumeric() {
            return false;
        }
    }
    true
}

fn generate_preview_content(pascal_case_name: &str) -> String {
    format!(
        "import React from 'react';
import {{ Input as FormilyInput }} from '@formily/antd-v5';
import {{ createBehavior, createResource }} from '@thevikalp/designable-core';
import {{ DnFC }} from '@thevikalp/designable-react';
import {{ createFieldSchema }} from '../Field';
import {{ AllSchemas }} from '../../schemas';
import {{ AllLocales }} from '../../locales';

export const {}: DnFC<React.ComponentProps<typeof FormilyInput>> = FormilyInput;

{}.Behavior = createBehavior({{
  name: '{}',
  extends: ['Field'],
  selector: (node) => node.props['x-component'] === '{}',
  designerProps: {{
    propsSchema: createFieldSchema(AllSchemas.{}),
  }},
  designerLocales: AllLocales.{},
}});

{}.Resource = createResource({{
  icon: '{}Source',
  elements: [
    {{
      componentName: 'Field',
      props: {{
        type: 'string',
        title: '{}',
        'x-decorator': 'FormItem',
        'x-component': '{}',
      }},
    }},
  ],
}});
",
        pascal_case_name,
        pascal_case_name,
        pascal_case_name,
        pascal_case_name,
        pascal_case_name,
        pascal_case_name,
        pascal_case_name,
        pascal_case_name,
        pascal_case_name,
        pascal_case_name
    )
}

fn generate_schema_content(pascal_case_name: &str) -> String {
    format!("import {{ ISchema }} from '@formily/react';

export const {}: ISchema = {{
  type: 'object',
  properties: {{
    placeholder: {{
      type: 'string',
      'x-decorator': 'FormItem',
      'x-component': 'Input',
      'x-component-props': {{
        placeholder: 'Enter placeholder text',
      }},
    }},
    maxLength: {{
      type: 'number',
      'x-decorator': 'FormItem',
      'x-component': 'NumberPicker',
      'x-component-props': {{
        placeholder: 'Maximum length',
      }},
    }},
    required: {{
      type: 'boolean',
      'x-decorator': 'FormItem',
      'x-component': 'Switch',
      'x-component-props': {{
        defaultChecked: false,
      }},
    }},
  }},
}};
", pascal_case_name)
}

fn generate_locale_content(pascal_case_name: &str) -> String {
    format!(
        "export const {} = {{
  'zh-CN': {{
    title: '{}',
    settings: {{
      'x-component-props': {{
        placeholder: '占位提示',
        maxLength: '最大长度',
      }},
    }},
  }},
  'en-US': {{
    title: '{}',
    settings: {{
      'x-component-props': {{
        placeholder: 'Placeholder',
        maxLength: 'Max Length',
      }},
    }},
  }},
  'ko-KR': {{
    title: '{}',
    settings: {{
      'x-component-props': {{
        placeholder: '플레이스홀더',
        maxLength: '최대 길이',
      }},
    }},
  }},
}};
",
        pascal_case_name,
        pascal_case_name,
        pascal_case_name,
        pascal_case_name
    )
}

fn update_index_file(index_path: &Path, export_line: &str) {
    let mut content = match fs::read_to_string(index_path) {
        Ok(c) => c,
        Err(e) => {
            eprintln!("❌ Error reading {}: {}", index_path.display(), e);
            return;
        }
    };

    if !content.contains(export_line) {
        content.push_str(&format!("\n{}", export_line));
        if let Err(e) = fs::write(index_path, content) {
            eprintln!("❌ Error writing {}: {}", index_path.display(), e);
            return;
        }
        println!("✅ Updated: {}", index_path.display());
    }
}
